NPEP-122 Tenancy API: Add Tenancy API design suggestions. (#178)
kubernetes-sigs/network-policy-api: npeps/npep-122.md

## API

### Tenant definition

We define Tenant as a set of namespaces.

> All Namespaces affected by Tenancy will be split into Tenants based on their label(s), we call these labels `tenancyLabels`. 

Author: I will update it a little, but the point is more about the semantics of tenancyLabels here than on API fields

    for `tenancyLabels`. To only select namespaces with existing labels extra `key: X, operator: Exists` condition should
    be added to the selector.

2. Joined with `tenancyLabels`. We select all namespace that all `tenancyLabels` present.

    **Cons**: applying all labels from `tenancyLabels` to a namespaces automatically makes it affected by tenancy.
    No way to make sure ALL namespaces in a cluster are a part of some tenant, e.g. newly created namespace will not
    be limited/protected by Tenancy rules until it gets `tenancyLabels` assigned, may potentially be a security problem.
    
    Example: by setting deny-from-other-tenants I want to ensure only namespaces with the same label `security-zone` value
    can send incoming connections. That policy will be violated if a new tenant is created without `security-zone` label,
    meaning it is not a part of any Tenant. This may potentially be solved by using allow-from-same-tenant + deny ALL
    ANP, but it is stricter, since it affects all incoming connections, not just from other pods.

### Peers and actions

> Based on the existing User Stories, Tenancy only needs Allow and Deny actions, and the only 2 types of peers are

Reviewer: I would disagree. I think user story 4.2 indicates that there is a need for PASS in case of strict tenant isolation, meaning "as a cluster admin I want to strictly disallow cross-tenant traffic, but owners of individual tenant namespaces can decide what they want to allow/deny in the scope of these namespaces"

Author: I am not sure I understand it correctly. So "strictly disallow cross-tenant traffic" means `Deny from/to notSameTenant` right? How would you combine it with PASS? maybe a yaml example would help me figure it out

Reviewer: What I mean is, instead of "deny from/to not same tenant", admin might want to define policies in the lines of "pass from/to same tenant, and deny everything else", especially when the egress story comes into play

Author: I think we need a new use case for that, the semantics is "as a cluster admin, I want to let namespaces decide what is allowed within their tenant, but strictly deny everything else" and it doesn't match any of the existing use stories
it also can be expressed as "deny from not same tenant" + "ANP PASS empty pod selector", but it is less obvious for a given use case
 @danwinship wdyt?

Reviewer 2: Tenancy-based rules are inherently east-west; you're always saying "traffic is allowed/denied between namespaces where these labels have the same/different values". You don't need to use a tenancy-based rule for N/S traffic, because the outside world can never be SameTenant.

IOW, if you want to implement "traffic is allowed within tenants and denied to/from everywhere else (including N/S)", then that would be:
1. a tenancy-based rule saying "traffic between tenants is denied"
2. a non-tenancy-based rule saying "N/S traffic to/from user namespaces is denied"

in which case there's still no need for a tenancy-based pass action.

----

Taking a step back, there are two concepts that we have sort of been mixing together:
1. "tenancy" - The idea that namespaces are grouped into tenants, and that you can use "same tenant" and "not same tenant" as selectors in policies.
2. "cluster default policy" - The idea that the administrator has some high-level plan for how networking works "normally" in the cluster (ie, when there aren't more-specific policies overriding it). eg:
    - "Traffic is permitted by default."
    - "Traffic within user namespaces is permitted by default, but traffic between user namespaces is blocked." 
    - "Traffic within the cluster is allowed by default, but cluster egress traffic is blocked." 

The concept of "cluster default policy" does not explicitly exist in ANP, but it's sort of lurking in the background, and all of the user stories for tenancy involve defining part of "cluster default policy", but tenancy isn't _all_ of "cluster default policy".

Part of the "tenancy is confusing / too powerful" problem that we're fixing now was because there was nothing stopping you from using tenancy-based rules in arbitrary ANPs/BANPs, even though the user stories only required tenancy-based rules in "cluster default policy" ANPs/BANPs... The new proposals below are better, since you can only define tenancy rules that apply to everyone, rather than randomly using a sameLabels rule inside a policy that only matches some traffic.

Reviewer: > IOW, if you want to implement "traffic is allowed within tenants and denied to/from everywhere else (including N/S)"

Well my point is, I was thinking the usecase as "I don't care about the traffic within tenants, but I want to make sure traffic is denied to/from everywhere else (including N/S)". This also is not a "cluster default policy" IMO because inside a tenant I do not want to define any rule, rather than "Traffic within user namespaces is permitted by default" which sounded like a BANP.

`SameTenant` and `NotSameTenant`.

Therefore, Tenancy rules will look something like:

```yaml
kind: NetworkTenancy
spec:
  tenancyLabels: ['label1']
  ingress:
    - actions: Deny
      from: NotSameTenant
    - actions: Allow
      from: SameTenant
  egress:
    - actions: Deny
      to: NotSameTenant
```

### Priorities

Based on User Story 4.4, we need to have Tenancy in the same priority range as ANP and BANP. There are multiple ways to do so:

1. Reuse ANP and BANP to define priority, link Tenancy API

```yaml
kind: AdminNetworkPolicy
spec:
  priority: 10
  # EITHER
  subject: <...>
  ingress: <...>
  egress: <...>
  # OR
  tenant: <tenant reference>
```

```yaml
kind: BaselineAdminNetworkPolicy
spec:
  # EITHER
  subject: <...>
  ingress: <...>
  egress: <...>
  # OR
  tenant: <tenant reference>
```

> We may want to allow either `subject`+`ingress`/`egress` or `tenant`, since having both at the same priority

Reviewer: ^ this

## API

### Tenant definition

For the purposes of this NPEP we define a Tenant as a set of namespaces.

> `tenancyLabels` is a set of label keys, based on which all Namespaces affected by Tenancy are be split into Tenants.

Reviewer 3: ```suggestion
`tenancyLabels` is a set of label keys, which are used to split selected Namespaces into Tenants.
```

A Tenant is identified by values of `tenancyLabels`, which are shared by all namespaces in a given Tenant.

There are 2 ways to select which namespaces should be affected by Tenancy rules:

1. Use a distinct new `namespaceSelector` field to define which namespaces should be affected by Tenancy, 
while the `tenancyLabels` field can be used to define how the selected namespaces are split into Tenants.

    **Cons**: selected namespace may not have some of the `tenancyLabels`, which will likely result in introducing "None" value
    for `tenancyLabels`. To only select namespaces with existing labels extra `key: X, operator: Exists` condition should
    be added to the selector.

```yaml
kind: NetworkTenancy
spec:

>   matchExpression:

Reviewer 3: Is there supposed to be a `namespaceSelector` in front of this?

```

```yaml
kind: NetworkTenancy
spec:
  priority: baseline
  ingress:
    - actions: Deny
      from: NotSameTenant
    - actions: Allow
      from: SameTenant
  egress:
    - actions: Deny
      to: NotSameTenant
```

### Another approach

There are 4 (6 if we consider Skip as an action, which I will do for now, but we can remove it later if we consider it
unnecessary) option for tenancy rules: `SameTenant: Allow/Skip/Deny`, `NotSameTenant: Allow/Skip/Deny`. 
From the user stories (and common sense) looks like the only useful combinations are `SameTenant: Allow/Skip` and `NotSameTenant: Deny`.
We can always replace `NotSameTenant: Deny` with Tenancy using `SameTenant: Allow/Skip` with lower priority 
and (B)ANP `Deny` policy for the same tenancy selector (or maybe even wider selector, e.g. all user pods).
This helps focus on which policy is defined inside one tenant, and then "everything else"/"other tenants"/etc
may be defined with existing CRDs.

In this case, we only need to define tenancy for the same tenant (which seems to be easier to understand).
Tenancy itself only needs a set of `tenancyLabels` and one of the `Allow` or `Skip` actions in every direction.

> If we consider tenancy to be always the highest-priority rule on ANP or BANP level, we could add one more field

Reviewer: Was this agreed in the user stories? I find it quite restrictive

- Users that don't care about tenancy can just ignore this CRD
- We can throw it away if we want to change API again

#### Final suggestion

Considering we agree with option 4 from the previous section on priorities specification, we can outline further details here.

For
```yaml
kind: NetworkTenancy
spec:
  tenancyLabels: ["user"]
  precedence: ANP/BANP
  action: Allow/Skip
```
there are only 3 useful combinations of precedence+action (BANP+Skip is the same as BANP+Allow), so we can express these
options in a different way, e.g. define 3 modes
- always-allow (ANP+Allow)
- delegate-to-NP (ANP+Skip)
- baseline-allow (BANP+Allow)

<details>
<summary>Full yaml examples (with the initial fields, will be updates as we agree on the final CRD format)</summary>

* 4.1 "overridable isolation"
```yaml
kind: NetworkTenancy
spec:
  tenancyLabels:
    - "user"
  precedence: BANP
  action: Allow
---
kind: BaselineAdminNetworkPolicy
spec:
  subject:
    namespaces:
      matchExpression:
        - key: user
          operator: Exists
  ingress:
    - action: Deny
      from:
        - namespaces: {}
  egress:
    - action: Deny
      to:
        - namespaces: {}

> ```

Author: first of all, I am not sure between-tenant behaviour is a necessary part of the tenancy. It is not obvious why "other tenants" should be different "other namespaces". Not sure if denying traffic for a namespace outside your tenant should be based on whether that namespace is also a part of some other tenant or not. 
Not saying it shouldn't be there, but the core concept of tenancy is `sameTenant`, which is also a dynamic selector that is not possible to express in any other way. Performance-wise I would probably even expect `notSameTenant` to be represented as a lower-priority deny rule for tenancy selector (with higher-priority `sameTenant` rule), as you don't need to create dynamic `notSameTenant` sets of pods for every tenant, but can only do that for `sameTenant`.
About including full use case (both allow and deny with their priorities): the main problem I faced is integration with ANP/BANP. When you only use `sameTenant,` you need only 1 implicit priority (for every use case we have highest-ANP and highest-BANP would work). But with deny rules, you may want deny as also highest-priority (use case 4.2) or lower than allow-from-monitoring (use case 4.4.1). Considering implicit priorities are adding more complexity, trying to track where exactly each one fits both for allow and deny rules just complicates the design (e.g. I am not sure how https://github.com/kubernetes-sigs/network-policy-api/pull/178#discussion_r1504764041 this example defined tenancy priority relative to ANP priority)

> ```

Reviewer 2: > first of all, I am not sure between-tenant behaviour is a necessary part of the tenancy.

Wow. OK, I guess I understand why we're having trouble agreeing on the details of the API now.

As I see it, tenancy is 100% about between-tenant behavior and 0% about anything else.

Look at the original user story:

> As a cluster admin, I want to build tenants in my cluster that are isolated from each other by default. Tenancy may be modeled as 1:1, where 1 tenant is mapped to a single Namespace, or 1:n, where a single tenant may own more than 1 Namespace.

It only talks about isolation between tenants, not anything that happens _inside_ the tenants. And the updated 4.1 says

> ... By default, cross-tenant traffic is dropped. ...

and again, neither 4.1 nor 4.2 says anything about what happens inside tenants.

The goal of tenancy is to put up walls between certain sets of namespaces. That's it. Traffic between tenants is restricted in a particular way, but all traffic that isn't between-tenant behaves exactly like it would have without the tenant definitions; you can make rules like "everybody can talk to the apiserver" or "namespaces with a user label can't talk to the internet by default", but you don't need a tenancy API to do that. You only need the tenancy API to specify between-tenant behavior.

> ```

Reviewer 4: Not sure if it helps, coming here from end-user/implementor perspective. The one single ask I have been getting always is: putting wall up between tenants (tenant here being set of namespaces); the users I know are designing almost always: "allow to youself, (isSame) deny to all other tenants(isNotSame)" => I know this is covered in one of the user stories already but re-hashing that that is the single most thing the current ANP/BANP can't do well enough; I think other things like allow {} to dns and deny {} can be done using anp/banp rules unless I am missing something.

- prio1 allow to your own tenant (I think this is 4.3 but ANP level? I have not yet heard a ask for BANP level)
- prio2 allow to DNS
- prio3 allow from monitoring
- last prio deny everything to everything total lockdown

the prio1 rule expression is what we struggle with on ANP because putting {} in the subject doesn't work as it does not allow reflecting tenancy in peers, rest can be done using anp/banp rules. so really the ask is almost always an easy way to say "please allow to my own tenant" because the deny is almost always going to be the default deny banp at the bottom which takes care of putting up walls. OR if you do the {} to {} deny in ANP that also solves that problem (I can see how the deny to other tenants can be useful if we can express that explicitly)

BANP can also be replaced with deny-all BANP
```yaml
kind: BaselineAdminNetworkPolicy
spec:
  subject:
    namespaces: {}
  ingress:
    - action: Deny
      from:
        - namespaces: {}
  egress:
    - action: Deny
      to:
        - namespaces: {}
```

* 4.2 strict isolation
```yaml
kind: NetworkTenancy
spec:
  tenancyLabels:
    - "user"
  precedence: ANP
  action: Skip
---
kind: AdminNetworkPolicy
spec:
  priority: 1
  subject:
    namespaces:
      matchExpression:
        - key: user
          operator: Exists
  ingress:
    - action: Deny
      from:
        - namespaces: {}
  egress:
    - action: Deny
      to:
        - namespaces: {}
```

* 4.3 Allow internal connections for tenants
```yaml
kind: NetworkTenancy
spec:
  tenancyLabels:
    - "user"
  precedence: ANP
  action: Allow
---
kind: BaselineAdminNetworkPolicy
spec:
  subject:
    namespaces:
      matchExpression:
        - key: user
          operator: Exists
  ingress:
    - action: Deny
      from:
        - namespaces: {}
  egress:
    - action: Deny
      to:
        - namespaces: {}

> ```

Author: if an admin allows namespaces to create their own network policies, then they may deny internal tenant connection by accident, and it will beak things. Then as an admin you may want to say always allow same tenant, which in turn lets namespaces specify policies for external-to-tenant traffic, but also ensures internal tenant connectivity is not broken by accident. (meaning it should prevent internal tenant connections problems both intentionally and by accident)

> ```

Reviewer 2: Yeah, I don't think that's a real use case. No admin who cares about security is going to want to _prohibit_ users from using NP within their own namespaces. (And any admin who doesn't care about security isn't going to be using ANP in the first place.)

* 4.4 Tenants interaction with (B)ANP
* 4.4.1 allow from monitoring + deny from not same tenant
```yaml
kind: NetworkTenancy
spec:
  tenancyLabels:
    - "user"
  precedence: ANP
  action: Skip
---
kind: AdminNetworkPolicy
spec:
  priority: 1
  subject:
    namespaces:
      matchExpression:
        - key: user
          operator: Exists
  ingress:
    - action: Allow
      from:
        - namespaces:
            namespaceSelector:
              matchLabels:
                kubernetes.io/metadata.name: monitoring-ns
    - action: Deny
      from:
        - namespaces: {}

> ```

Author: I am not sure in the user stories context, we ever meant "all namespaces" as explicitly all namespaces in the cluster even if they are not a part of tenancy (like why would we discuss non-tenancy namespace here in the use cases). Not sure if the wording is wrong, or we have been thinking about these use cases differently this whole time. e.g. I interpret 4.4.1 as "I want to allow ingress traffic from the monitoring namespace to all (tenancy-affected) namespaces." as you can see in the yaml example. (I guess having yamls helps clarify this)
the same applies to the next comment.

> ```

Reviewer 2: Looking at this again:
- If the precedence is "ANP, Tenancy, NP, BANP" then the tenancy rule is ignored, because the ANP's Deny would run first
- If the precedence is "Tenancy, ANP, NP, BANP" then the Allow-from-monitoring rule is ignored, because the Tenancy's Skip rule would bypass the ANP.

Author: In this example it works as your second option, but Skip action is only applied to sameTenant namespaces, while monitoring is **not** sameTenant. So traffic from monitoring ns won't match tenancy rules (unless there is a tenant of which monitoring is a part, then monitoring will be passed as sameTenant)

  ingress:
    - action: Allow
      from:
        - SameTenant
        - pods:
            podSelector: {}
```
**CONS** 
- "switch" that enables some peer types is not the best API design
- gives more flexibility than intended (multiple tenancy definitions)
- conflicts with singleton BANP, meaning that if Tenancy is defined on BANP level, general-purpose BANP selecting different
pods can't be created.

2. Reuse ANP and BANP to define priority, insert Tenancy definition and Rules to `ingress` and `egress`.

It splits the tenancy labels and namespace selector which defines the namespace that are affected by the tenancy.

```yaml
kind: AdminNetworkPolicy
spec:
  priority: 10
  subject:
    namespaces:
      matchLabels:
        kubernetes.io/metadata.name: sensitive-ns
  ingress:
    - action: Allow
      from:
      - tenant:
          labels: ["user"]
          tenant: Same
```

**CONS** complicates selection or affected namespaces, allows strange configuration, where namespaces selected by 
`subject` have no tenancy labels. Has the same problem as the original design.

3. Create 2 objects with ANP and BANP priorities (let's say NetworkTenancy and BaselineNetworkTenancy)

```yaml

> kind: NetworkTenancy

Author: Good point, we actually had a working name of `TenancyNetworkPolicy` (which I haven't added to the NPEP yet, but now I will), lmk if that sounds good.
About the future use cases: we are usually trying to not predict future use cases and not over-engineer APIs before it is needed. So if we need more similar functionality at some point, we probably will create a new API and some migration tool, if that makes sense.

spec:
  priority: 10
  tenancyLabels: ["user"]
  action: SkipSameTenant
---
kind: BaselineNetworkTenancy
spec:
  priority: 10
  tenancyLabels: ["user"]
  action: SkipSameTenant
```

While multiple ANPs with the same priority are allowed, we probably can allow multiple Tenancies or Tenancy and ANP
with the same priority, but if we decide to only allow ANP per priority, Tenancy needs to be accounted for in the same range.

> **CONS**: BANP doesn't have a priority, to use this method we would need to define a priority for BANP.

Author: I think you are talking about option 4.1:  Create 2 new objects with implicit priorities. (I will add that one too)
The main "con" for it is that it will be 2 new CRDs with just 2 fields, but following the existing structure is the "pro", so let's discuss which one makes more sense

4. Create 1 new object with implicit priorities.

> `precedence` field + reserved highest-priority rule before (B)ANP

Author: skip action actually is made to allow NP "override" something that ANP can't. We explain this action sometimes as "skip=delegate to namespaced network policies". It is useful when you **want** to skip some of the following ANP rules, but you don't want to make != exceptions for e.g. same-namespace traffic. Also, cluster admin doesn't always have a decision (allow or deny) for some traffic, and wants to let network policies decide that.
As per example, if I understand it correctly, you don't need to include namespaces you don't want to be affected by tenancy to the tenancy policy (not sure if there is any other need to override tenancy policy)

## API

### Tenant definition

For the purposes of this NPEP we define a Tenant as a set of namespaces.
`tenancyLabels` is a set of label keys, based on which all Namespaces affected by Tenancy are be split into Tenants.
A Tenant is identified by values of `tenancyLabels`, which are shared by all namespaces in a given Tenant.

There are 2 ways to select which namespaces should be affected by Tenancy rules:

1. Use a distinct new `namespaceSelector` field to define which namespaces should be affected by Tenancy, 
while the `tenancyLabels` field can be used to define how the selected namespaces are split into Tenants.

    **Cons**: selected namespace may not have some of the `tenancyLabels`, which will likely result in introducing "None" value
    for `tenancyLabels`. It brings more complexity and is not required for any of the listed use cases.

```yaml
spec:
  matchExpression:
    - key: system-namespace
      operator: DoesNotExist
    - key: user
      operator: Exists
  tenancyLabels:
    - user
```

2. Overload `tenancyLabels` and implicitly apply tenancy rules only to namespaces where `tenancyLabels` are present.

>     **Cons**: the simplest option that covers all given use cases.

Author: this doesn't allow using `kubernetes.io/metadata.name` label, which is a natural choice for allow-same-namespace policy, as it means all namespaces will be selected by tenancy

    namespaces:
      matchExpression:
        - key: user
          operator: Exists
  ingress:
    - action: Deny
      from:
        - namespaces: {}
  egress:
    - action: Deny
      to:
        - namespaces: {}
```

* 4.4 Tenants interaction with (B)ANP
* 4.4.1 allow from monitoring + deny from not same tenant
```yaml
kind: TenancyNetworkPolicy
spec:
  tenancyLabels:
    - "user"
  precedence: ANP
  action: PassSameTenant
---
kind: AdminNetworkPolicy
spec:
  priority: 1
  subject:
    namespaces:
      matchExpression:
        - key: user
          operator: Exists
  ingress:
    - action: Allow
      from:
        - namespaces:
            namespaceSelector:
              matchLabels:
                kubernetes.io/metadata.name: monitoring-ns
    - action: Deny
      from:
        - namespaces: {}
```
* 4.4.2 allow from same tenant + BANP deny all
```yaml
kind: TenancyNetworkPolicy
spec:
  tenancyLabels:
    - "user"
  precedence: BANP
  action: PassSameTenant
---
kind: BaselineAdminNetworkPolicy
spec:
  subject:
    namespaces:
      matchExpression:
        - key: user
          operator: Exists
  ingress:
    - action: Deny
      from:
        - namespaces: {}
```

> </details>

Reviewer 4: all the examples provided here are priceless and well thought out, thank you Nadia!!
I have a silly question:
whichever way we decide to do, If I do:
 tenancyLabels:
    - "user" (nsA:prod, nsB:staging, nsC:prod, nsD: staging, nsE:prod)
    - "color" (nsA:red, nsB:red, nsC:blue, nsD:blue, nsE:red)
   
So subject is all namespaces that have either of these tenancyLabels set right ? (not ALL?).

then the concept of `sameTenant` means any other tenant that has `Either of these 3 keys with the same value` OR `any tenant that has ALL of these 3 keys with the same value`; is it && or || ?

so who is sameTenant here for nsA? is it nsC, nsB and nsE? or only nsE ?

Author: We have mostly discussed this for one label case, but I think it applies to multiple labels too (will add a section for that)
So for one label, we decided that only namespaces that have that label will be selected to avoid potential None tenant. Therefore for multiple labels, only namespaces that have all tenancy labels will be selected for the same reason. So sameTenant for nsA is namespace with user=prod AND color=red.
